### PR TITLE
Don't show buildings on minimap that are not visible in main view

### LINKED ISSUE #3778: Minimap reveals unseen enemy buildings
In this screenshot, the minimap shows that the blue player has a building near the screen center. No such building is visible.

The debug console reveals that the building in question is a port which is apparently located partly in seen and partly in unexplored terrain.
Either this building should be shown in the main view as well as on the minimap, or the minimap should not show buildings invisible in the main view.

[Savegame.wgf.zip](https://github.com/widelands/widelands/files/4339510/Savegame.wgf.zip)

### PR DESCRIPTION
**Pull request by Niektory**
Fixes #3778 

Works but is a little ugly and may be buggy - I copy-pasted some code without fully understanding it.

### CODE REVIEW

File: src/graphic/minimap_renderer.cc [hunk 1]
```diff
@@ -45,11 +45,30 @@ int round_up_to_nearest_even(int number) {
...
+	const int32_t mapwidth = map.get_width();
+	Widelands::FCoords f = map.get_fcoords(coords);
+	Widelands::MapIndex i = Widelands::Map::get_index(f, mapwidth);
+	move_r(mapwidth, f, i);
+
+	if (player == nullptr || player->see_all()) {
+		return 2;
+	} else if (player != nullptr) {
+		const auto& field = player->fields()[i];
+		return field.vision;
```
Comment (Noordfrees) [inline src/graphic/minimap_renderer.cc]:
```suggestion
		return player->fields()[i].vision;
```

File: src/graphic/minimap_renderer.cc [hunk 2]
```diff
@@ -45,11 +45,30 @@ int round_up_to_nearest_even(int number) {
...
 }
 
+uint16_t check_vision(const Widelands::Coords& coords,
+					  const Widelands::Player* player,
+					  const Widelands::EditorGameBase& egbase) {
+	const Widelands::Map& map = egbase.map();
+	const int32_t mapwidth = map.get_width();
+	Widelands::FCoords f = map.get_fcoords(coords);
+	Widelands::MapIndex i = Widelands::Map::get_index(f, mapwidth);
+	move_r(mapwidth, f, i);
```
Comment (Noordfrees) [inline src/graphic/minimap_renderer.cc]: Please add a comment explaining what this does

File: src/graphic/minimap_renderer.cc [hunk 3]
```diff
@@ -75,7 +94,8 @@ inline RGBColor calc_minimap_color(const Widelands::EditorGameBase& egbase,
 
 			if (((layers & MiniMapLayer::Flag) && dynamic_cast<Flag const*>(immovable)) ||
 			    ((layers & MiniMapLayer::Building) &&
```
Comment (Noordfrees) [inline src/graphic/minimap_renderer.cc]:
```suggestion
			if (((layers & MiniMapLayer::Flag) && immovable->descr().type() == Widelands::MapObjectType::FLAG)) ||
			    ((layers & MiniMapLayer::Building) &&
```

File: src/graphic/minimap_renderer.cc [hunk 4]
```diff
@@ -75,7 +94,8 @@ inline RGBColor calc_minimap_color(const Widelands::EditorGameBase& egbase,
...
-			     dynamic_cast<Widelands::Building const*>(immovable))) {
+			     dynamic_cast<Widelands::Building const*>(immovable) &&
```
Comment (Noordfrees) [inline src/graphic/minimap_renderer.cc]:
```suggestion
			     immovable->descr().type() >= Widelands::MapObjectType::BUILDING &&
```

**Review by Noordfrees — COMMENTED** [review]
Tested and working, some performance suggestions inline.

This is for b22.

File: src/graphic/minimap_renderer.cc (hunk 2 above)
Comment (Niektory) [reply src/graphic/minimap_renderer.cc]: I copy-pasted this line from `do_draw_minimap()`, and testing shows it's wrong here. Removing...

**Comment by Niektory** [comment]
I also tried doing it the other way around - showing the buildings in the main view - since that would be the better way I believe. The "currently partially visible" case is not too bad, but the "formerly partially visible" case is tricky. I see two ways to handle it:

- Store all positions in `Player::Field`. That would also require storing something to distinguish them. Currently it stores only what type of immovable was seen on a node, but not which part of it, so if we store all the positions seen we can't tell what's the primary position.
- Always store the primary position in `Player::Field` - even in unseen nodes - when any of the positions is seen. Don't store secondary positions.

**Comment by Noordfrees** [comment]
Lots of instances of `Player::Field` are allocated in big games so it needs to use as little memory as possible. So storing the `MapIndex` of the MapObject's main Coords in addition to the `Descr*` would be best imho.

**Comment by gunchleoc** [comment]
When a building ranges over multiple fields, I think it would be precise enough for the minimap to draw it on any one of these fields and it doesn't matter which one. So, the minimap could remember the serials of the buildings already drawn.

**Comment by Niektory** [comment]
There will be conflicts with #4046 so I'll wait until that's merged.

**Comment by Niektory** [comment]
There will be conflicts with #4046 so I'll wait until that's merged.

**Pull request closed without merging**